### PR TITLE
pagmo2, pythonPackages.pygmo fixes

### PR DESCRIPTION
**Pull request by worldofpeace**
###### Motivation for this change
Fix #51399 with @dotlambda [change](https://github.com/dotlambda/nixpkgs/commit/4d0a5619cd8e54e4572c1a6d26b636bacd139d46) which looked fine.

This change was meant to fix issues with https://github.com/NixOS/nixpkgs/pull/50645 so I've brought it here as well for testing.

###### Things done

I've checked that at least that the packages in question would build.

---

### CODE REVIEW

File: pkgs/development/python-modules/ase/default.nix [hunk 1]
```diff
@@ -10,12 +10,12 @@
 }:
 
 buildPythonPackage rec {
-  version = "3.16.2";
+  version = "3.17.0";
   pname = "ase";
 
   src = fetchurl {
```
Comment (dotlambda) [inline pkgs/development/python-modules/ase/default.nix]:
```suggestion
  src = fetchFromGitLab {
```

File: pkgs/development/python-modules/pygmo/default.nix [hunk 2]
```diff
@@ -11,38 +11,44 @@
...
-  propagatedBuildInputs = [ numpy cloudpickle ipyparallel numba ];
+  src = pagmo2WithPython;
 
   preBuild = ''
-    cp -v -r $src/* .
-    cmake -DCMAKE_INSTALL_PREFIX=$out -DPAGMO_BUILD_TESTS=no -DCMAKE_SYSTEM_NAME=Linux -DPagmo_DIR=${pagmo2} -DPAGMO_BUILD_PYGMO=yes -DPAGMO_BUILD_PAGMO=no -DPAGMO_WITH_EIGEN3=yes -DPAGMO_WITH_NLOPT=yes -DNLOPT_LIBRARY=${nlopt}/lib/libnlopt_cxx.so -DPAGMO_WITH_IPOPT=yes -DIPOPT=${ipopt}
-
-    make install
-    mv $out/lib/python*/site-packages/pygmo wheel
+    mv ${python.sitePackages}/pygmo wheel
```
Comment (dotlambda) [inline pkgs/development/python-modules/pygmo/default.nix]: :+1:

**Comment by dotlambda** [comment]
Just because this might of interest to some:
```
$ nix why-depends -f . python3.pkgs.pygmo pagmo2
/nix/store/j2k5f520hj79njqp1kb6fxhnqxx2kjsz-python3.7-pygmo-2.9
╚═══lib/python3.7/site-packages/pygmo/core.so: …volume objects......./nix/store/asyfspmrvghxq5v0fyzw2f2r4yzjbv89-pagmo2-2.9/include/pagmo/popula…
    => /nix/store/asyfspmrvghxq5v0fyzw2f2r4yzjbv89-pagmo2-2.9
$ nix why-depends -f . python3.pkgs.pygmo python3.pkgs.pygmo.src
'python3.pkgs.pygmo' does not depend on 'python3.pkgs.pygmo.src'
```
This is exactly what we expect as the relevant output of `pagmo2WithPython` is copied to `pygmo`'s output.

File: pkgs/development/python-modules/ase/default.nix (hunk 1 above)
Comment (dotlambda) [reply pkgs/development/python-modules/ase/default.nix]: Actually, `fetchPypi` is even better.

File: pkgs/development/python-modules/pygmo/default.nix [hunk 3]
```diff
@@ -11,38 +11,44 @@
...
+let
+  propagatedBuildInputs = [ numpy cloudpickle ipyparallel numba ];
+
+  pagmo2WithPython = pagmo2.overrideAttrs (oldAttrs: {
+    cmakeFlags = oldAttrs.cmakeFlags ++ [
+      "-DPAGMO_BUILD_PYGMO=yes"
+      "-DPAGMO_BUILD_PAGMO=no"
+      "-DPagmo_DIR=${pagmo2}"
+    ];
+    buildInputs = [ cmake eigen nlopt ipopt boost pagmo2 ] ++ propagatedBuildInputs;
```
Comment (Mic92) [inline pkgs/development/python-modules/pygmo/default.nix]: `cmake` belongs to nativeBuildInputs.

**Comment by worldofpeace** [comment]
@dotlambda Sorry I was away for a bit, I'll fix this up and it should be good :+1:

**Comment by dotlambda** [comment]
I would have done it myself but it seems like you disallowed "edits by maintainers".

**Comment by worldofpeace** [comment]
> I would have done it myself but it seems like you disallowed "edits by maintainers".

Oops, didn't mean to do that :rofl:

File: pkgs/development/python-modules/pygmo/default.nix [hunk 4]
```diff
@@ -11,38 +11,44 @@
 , cloudpickle
 , ipyparallel
 , numba
+, python
 }:
```
Comment (dotlambda) [inline pkgs/development/python-modules/pygmo/default.nix]: You should remove `cmake` from above here.

**Comment by worldofpeace** [comment]
Lol ryantm opened #50645 Nov 18 and it's Dec 18

Good to see this finally hit :+1:

**Pull request merged**